### ISSUE TITLE
Can this be used for classical ensembling? Intersection of tokenizers? 

### ISSUE THREAD

**Issue by jukofyork** [cg123/mergekit#106]
Effectively I want to run multiple models unaltered right up to the final softmax layer and then take a weighted sum of the of the pre-softmax inputs. This is mathematically equivalent to taking the weighted geometric mean of the softmax outputs / probabilities.

This is more classical ensembling than merging and takes more compute to run, but I think it could be very useful for a few different reasons:

- You should be able to take a base model and it official chat/instruction tuned model and get back some of the extra performance of the base model and/or remove some of the censorship, at the cost of possibly a more unruly output which could be balanced empirically via the weighting factor. 
-  It would also work as a method of regularisation for some of the massively overfit fine-tuned models people are making on huggingface, by again allowing you to regularise back towards the base model using the weighting factor. 
- You should be able to take several different models that say used codellama as the base model and then ensemble them to reduce the variance, and hopefully get a better performing model overall (with 2x 34b codellamas being no worse to run than a single 70b and n=2 theoretically giving the best variance reduction to extra compute ratio, etc). People are also starting to fine-tune deepseek-coder now.

I've been looking at the llama.cpp "Speculative Decoding" example and think it wouldn't be that hard to adapt it to do this, but just wondered if this might already be possible or easy to implement using MergeKit?

I also see you have the option for a union of tokenizers, but the intersection of tokenizers might also be interesting/useful. In terms of the above this is just assuming the geometric mean of zero with anything else is zero. Some of the Chinese models with lots of Unicode tokens could possibly even be improved on non-Chinese related texts by constraining away these tokens, but I guess it depends on how the tokenizers originally broke up longer words.

**Comment by jukofyork** [comment]
I've just found the mixtral_moe.py script and wonder if this could be  adapted to do create vanilla ensembles, but I'm not sure how the gate works for these new type of MOEs. 

In the original formulation of MOE it was implemented via a sigmoid output and a scaler product, but the fact that the real Mixtral runs faster (using only 2 of 8 experts, IIRC) makes me think the sigmoid is getting thresholded somewhere?

If it wasn't then the weighted sum could easily be implimented by making a dumb/constant "expert" who's gating parameter is set using the inverse-logistic / log-odds value for the weighting we want.

**Comment by jukofyork** [comment]
I've been looking sole more and it appears this technique is called "late fusion":

https://aclanthology.org/2023.findings-emnlp.430.pdf

This paper is using it for 3-level sentiment analysis so not quite the same as described above, but it seems to be the first reference to this exact term (both the 2020 and 2020 papers it links the name "late fusion" to don't actually use the "late fusion" term).

**Comment by cg123** [comment]
This is something I definitely want to play with at some point. The `mergekit-moe` script won't be directly applicable, as it leans pretty heavily on being able to shove model weights into the Mixtral architecture - their routing is a linear transformation of hidden state -> softmax -> top K, and it'd be hard to wrangle that into a concrete weighting. To do things properly custom inference code would be needed for the resulting model. 

You're right, a tokenizer intersection would be pretty useful both for this kind of ensembling and for general model manipulation - I'll spin out a little script to do that in the next couple of days.

Thanks for your thoughts on this!

**Comment by jukofyork** [comment]
@cg123 Yeah, I think it would be quite interesting to see what effect ensembling would have. I get the impression that a lot of the fine-tuning that is going on is crying out for some kind of regularization as the datasets people are using are so tiny in comparison to what was actually used to train the base models in the first place.

I'd be quite interested just to see the stats on the intersection of tokenizers even if it doesn't really work - watching Goliath-120B write out stories *very* slowly gives you a good chance to see just how it tokenizes the longer words, and it would be interesting to see how differently some of the Chinese 70B models tokenize considering they have so many more Unicode tokens, etc.

**Comment by jukofyork** [comment]
I'm going to close this too, as after playing around a lot today: it's pretty clear this idea simply won't work unless the models both use *exactly* the same prompt format.

I did manage to hack the MOE script to use `torch.zeros()` in place of `randn()`, and it did make a "sort of ensemble" and should in theory have been different to plain weight merging from the swish activations after the gating:

https://cdn-uploads.huggingface.co/production/uploads/65921b0435c41262d6b68b34/-scdQY3VUk19RsiULSo-7.png

But it basically sucked and the different prompt formats and stop-tokens seemed to just clash badly and even when it did work, it wasn't really any better than the base model.

Also, this was still using the main donor transformer blocks so any "late merging" is likely to be even worse.
